api/src/wfl/service/slack.clj

```diff
@@ -38,30 +35,41 @@
   [url description]
   (format "<%s|%s>" url description))
 
-(defn ^:private slack-api-raise-for-status
-  "Slack API has its own way of reporting
-   statuses, so we need to parse the `body`
-   to raise for status."
-  [body]
-  (let [response (json/read-str body)]
-    (when-not (response "ok")
-      (throw (ex-info "failed to notify via Slack"
-                      {:error (response "error")})))
+;; More information on the meaning of error responses:
+;; https://api.slack.com/methods/chat.postMessage#errors
+;;
+(defn ^:private post-message
+  "Post `message` to `channel`."
+  [channel message]
+  (let [headers {:Authorization (str "Bearer " (env/getenv "WFL_SLACK_TOKEN"))}
+        body    (json/write-str {:channel channel :text message})]
+    (-> "https://slack.com/api/chat.postMessage"
+        (http/post {:headers      headers
+                    :content-type :application/json
+                    :body         body})
+        util/response-body-json)))
+
+;; Slack API has its own way of reporting statuses:
+;; https://api.slack.com/web#slack-web-api__evaluating-responses
+;;
+(defn ^:private post-message-and-throw-if-failure-impl
+  "Post `message` to `channel` and throw if response indicates a failure."
+  [channel message]
+  {:pre [(valid-channel-id? channel)]}
+  (let [response (post-message channel message)]
+    (when-not (:ok response)
+      (throw (ex-info "Slack API chat.postMessage failed"
+                      {:channel  channel
+                       :message  message
+                       :response response})))
     response))
 
-(defn post-message
-  "Post message to channel and pass response to callback."
-  ([channel message]
-   {:pre [(valid-channel-id? channel)]}
-   (-> "https://slack.com/api/chat.postMessage"
-       (http/post {:headers      {:Authorization (str "Bearer " @token)}
-                   :content-type :application/json
-                   :body         (json/write-str {:channel channel
-                                                  :text    message})})
-       :body
-       slack-api-raise-for-status))
-  ([channel message callback]
-   (callback (post-message channel message))))
+;; Deferring implementation to allow for response verification in tests.
+;;
+(defn ^:private post-message-and-throw-if-failure
+  "Post `message` to `channel` and throw if response indicates a failure."
+  [channel message]
+  (post-message-and-throw-if-failure-impl channel message))
```

Comment:
Does the top-level comment help?

> Deferring implementation to allow for response verification in tests.

What I really wanted was an option to mock `post-message-and-throw-if-failure`, calling the original method, verifying that the Slack message was sent, then returning what it would have returned.  Happy to chat more about this.

Reply:
Thanks for pulling the thread.